src/auditwheel/mod.rs

```diff
@@ -0,0 +1,249 @@
+                versions,
+            });
+        }
+    }
+    Ok(symbols)
+}
+
+/// Find incompliant symbols from symbol versions
+fn find_incompliant_symbols(
+    elf: &Elf,
+    symbol_versions: &[String],
+) -> Result<Vec<String>, AuditWheelError> {
+    let mut symbols = Vec::new();
+    let strtab = &elf.strtab;
+    for sym in &elf.syms {
+        if sym.st_type() == STT_FUNC {
+            let name = strtab
+                .get(sym.st_name)
+                .unwrap_or(Ok("BAD NAME"))
+                .map_err(AuditWheelError::GoblinError)?;
+            for symbol_version in symbol_versions {
+                if name.ends_with(&format!("@{}", symbol_version)) {
+                    symbols.push(name.to_string());
+                }
+            }
+        }
+    }
+    Ok(symbols)
+}
+
+/// An (incomplete) reimplementation of auditwheel, which checks elf files for
+/// manylinux compliance. Returns an error for non compliant elf files
+///
+/// Only checks for the libraries marked as NEEDED, but not for symbol versions
+/// (e.g. requiring a too recent glibc isn't caught).
+pub fn auditwheel_rs(
+    path: &Path,
+    target: &Target,
+    manylinux: &Manylinux,
+) -> Result<(), AuditWheelError> {
+    if !target.is_linux() || matches!(manylinux, Manylinux::Off) {
+        return Ok(());
+    }
+    let policy = POLICIES
+        .iter()
+        .find(|p| p.name == manylinux.to_string())
+        .unwrap();
+    let mut file = File::open(path).map_err(AuditWheelError::IOError)?;
+    let mut buffer = Vec::new();
+    file.read_to_end(&mut buffer)
+        .map_err(AuditWheelError::IOError)?;
+    let elf = Elf::parse(&buffer).map_err(AuditWheelError::GoblinError)?;
+    // This returns essentially the same as ldd
+    let deps: Vec<String> = elf.libraries.iter().map(ToString::to_string).collect();
+    let versioned_libraries = find_versioned_libraries(&elf, &buffer)?;
+
+    let mut offenders = HashSet::new();
+    for dep in deps {
+        // Skip dynamic linker/loader
+        if dep.starts_with("ld-linux") || dep == "ld64.so.2" || dep == "ld64.so.1" {
+            continue;
+        }
+        if !policy.lib_whitelist.contains(&dep) {
+            offenders.insert(dep);
+        }
+    }
+    for library in versioned_libraries {
+        if !policy.lib_whitelist.contains(&library.name) {
+            offenders.insert(library.name.clone());
+            continue;
+        }
+        let mut versions: HashMap<String, HashSet<String>> = HashMap::new();
+        for v in &library.versions {
+            let mut parts = v.splitn(2, '_');
+            let name = parts.next().unwrap();
+            let version = parts.next().unwrap();
+            versions
+                .entry(name.to_string())
+                .or_default()
+                .insert(version.to_string());
+        }
+        let arch = target.target_arch().to_string();
+        let arch_versions = &policy.symbol_versions[&arch];
```

Comment:
Does this panic when using an unsupported? In that case it would be better to use `.get` and error appropriately